src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,596 @@
+{
+    Poco::File dir_handle(local_path_);
+    Poco::File data_file(local_path_ + "/data.bin");
+    Poco::File meta_file(local_path_ + "/meta.txt");
+    if (!dir_handle.exists() || !data_file.exists() || !meta_file.exists())
+    {
+        LOG_ERROR(&Poco::Logger::get("RemoteCacheController"), "not exists directory" + local_path_);
+        return nullptr;
+    }
+
+    std::ifstream meta_fs(local_path_ + "/meta.txt");
+    Poco::JSON::Parser meta_parser;
+    auto meta_jobj = meta_parser.parse(meta_fs).extract<Poco::JSON::Object::Ptr>();
+    auto remote_path = meta_jobj->get("remote_path").convert<std::string>();
+    auto schema = meta_jobj->get("schema").convert<std::string>();
+    auto cluster = meta_jobj->get("cluster").convert<std::string>();
+    auto downloaded = meta_jobj->get("downloaded").convert<std::string>();
+    auto mod_ts = meta_jobj->get("last_mod_ts").convert<UInt64>();
+    if (downloaded == "false")
+    {
+        LOG_ERROR(&Poco::Logger::get("RemoteCacheController"), "not a downloaded file " + local_path_);
+        return nullptr;
+    }
+    auto size = data_file.getSize();
+
+    auto cntrl = std::make_shared<RemoteCacheController>(schema, cluster, remote_path, mod_ts, local_path_, nullptr, finish_callback);
+    cntrl->download_finished = true;
+    cntrl->current_offset = size;
+    meta_fs.close();
+
+    finish_callback(cntrl.get());
+    return cntrl;
+}
+
+RemoteCacheController::RemoteCacheController(
+    const std::string & schema_,
+    const std::string & cluster_,
+    const std::string & path_,
+    UInt64 mod_ts_,
+    const std::string & local_path_,
+    std::shared_ptr<ReadBuffer> readbuffer_,
+    std::function<void(RemoteCacheController *)> const & finish_callback)
+{
+    download_thread = nullptr;
+    schema = schema_;
+    cluster = cluster_;
+    local_path = local_path_;
+    remote_path = path_;
+    last_mod_ts = mod_ts_;
+    valid = true;
+    if (readbuffer_ != nullptr)
+    {
+        download_finished = false;
+        current_offset = 0;
+        remote_readbuffer = readbuffer_;
+        // setup local files
+        out_file = new Poco::FileOutputStream(local_path_ + "/data.bin", std::ios::out | std::ios::binary);
+        out_file->flush();
+
+        Poco::JSON::Object jobj;
+        jobj.set("schema", schema_);
+        jobj.set("cluster", cluster_);
+        jobj.set("remote_path", path_);
+        jobj.set("downloaded", "false");
+        jobj.set("last_mod_ts", mod_ts_);
+        std::stringstream buf; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+        jobj.stringify(buf);
+        Poco::FileOutputStream meta_file(local_path_ + "/meta.txt", std::ios::out);
+        meta_file.write(buf.str().c_str(), buf.str().size());
+        meta_file.close();
+
+        backgroupDownload(finish_callback);
+    }
+}
+
+int RemoteCacheController::waitMoreData(size_t start_offset_, size_t end_offset_)
```

Comment:
What about cache eviction? Do you plan something about it?

Reply:
cached files are evicted in RemoteReadBufferCache::clearLocalCache()
the eviction will be run when the total size of cached files is larger then the limitation.